[fix][broker] Execute per-topic entry filters with the same classloader

nicoloboschi opened the pull request:
Fixes #19361

### Motivation

Currently the entry filters per-topic/per-namespace are recreated from scratch (search in fs, unzip nar, classloader creation) every time a topic with entry filters enabled is created/updated.
This leads to multiple issues:
1. The topic creation path is affected by the time for looking up the file system and the unpacking (cpu intensive) 
2. Each topic has his own `NarClassLoader` instance that has a relevant memory footprint impact

 


### Modifications
1. Made the `EntryFilterProvider` a stateful class. In the constructor all the entry filter definitions and NarClassLoader are loaded, together with the broker entry filters
2. When the topic is created, using the same NarClassLoader a new EntryFilterWithClassLoader instance is created.



One different thing from before is that now the same entry filter class is loaded once. This means if a entry filter relies on `static` variables, those point to the same memory address for each EntryFilter invocation, even for different topics. 
I think it's acceptable because:
1. This feature has never been released so it's not a real breaking change
2. The behaviour of EntryFilter is consistent with the broker ones. For the developer this means to not have to think to different contexts based on how the filter is deployed.


Other things in the pull request:
1. Moved the entryFilter update in `AbstractTopic` to the `initialize` method
2. Improved the entry filters override logic where, if the broker denied to override the broker entry filters but not filters were configured, then the topic's filter was used. This is wrong since the Pulsar admin doesn't expect any filter to be executed if `allowOverrideEntryFilters` is false
3. Added test about execution ordering of entry filters, if more than one. This is very important since it can cause performance regression based on the chain of entry filters configured. 

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. Please attach the local preview screenshots (run `sh start.sh` at `pulsar/site2/website`) to your PR description, or else your PR might not get merged. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

### Matching PR in forked repository

lhotari commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/plugin/EntryFilterProvider.java:
```diff
@@ -37,74 +43,78 @@
+        }
+    }
+
+    public List<EntryFilterWithClassLoader> loadEntryFiltersForPolicy(EntryFilters policy)
```
In the original implementation, `EntryFilterWithClassLoader` is unnecessarily used as a top level concept. I think that this is a bad approach. It's sufficient to use `EntryFilter`. 

Please refactor the solution to hide `EntryFilterWithClassLoader`. It's possible that it could be removed completely.

eolivelli replied:
Unfortunately we need to keep a reference to the NarClassLoader, this is a pattern we have in other plugins that can be loaded from a Nar file

lhotari reviewed (changes requested):
Good work @nicoloboschi . I have a comment about `EntryFilterWithClassLoader`.

lhotari commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:
```diff
@@ -463,4 +472,93 @@ private void verifyBacklog(String topic, String subscription,
+            producer.send("test");
+        }
+
+        EntryFilterWithClassLoader mockFilterReject = mock(EntryFilterWithClassLoader.class);
+        when(mockFilterReject.filterEntry(any(Entry.class), any(FilterContext.class))).thenReturn(
+                EntryFilter.FilterResult.REJECT);
+        EntryFilterWithClassLoader mockFilterAccept = mock(EntryFilterWithClassLoader.class);
+        when(mockFilterAccept.filterEntry(any(Entry.class), any(FilterContext.class))).thenReturn(
+                EntryFilter.FilterResult.ACCEPT);
+        if (overrideBrokerEntryFilters) {
+            setMockFilterToTopic((PersistentTopic) pulsar.getBrokerService()
+                    .getTopicReference(topic).get(), List.of(mockFilterReject, mockFilterAccept));
+        } else {
+            setMockFilterToTopic((PersistentTopic) pulsar.getBrokerService()
+                    .getTopicReference(topic).get(), List.of());
+            setMockBrokerFilter(List.of(mockFilterReject, mockFilterAccept));
+        }
```
I think that we need to get rid of this type of mocking. It's not that this PR has introduced this approach. 

It would be useful it the EntryFilterProvider could be overridden in tests. It becomes easier if EntryFilterProvider is a simple interface and the existing EntryFilterProvider is renamed to EntryFilterProviderImpl or DefaultEntryFilterProvider. 

Instead of injecting fields and doing other hacks, filters could simply be tested by passing an EntryFilterProvider in tests.

lhotari commented:
> Currently the entry filters per-topic/per-namespace are recreated from scratch (search in fs, unzip nar, classloader creation) every time a topic with entry filters enabled is created/updated.

One clarification to this: this applies to all entry filters in the master branch. The master branch has conflicting changes, but it will also load per-broker filters using the same logic as per-topic filters. The filters just won't get used at all unless allowOverrideEntryFilters=true.

lhotari commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/service/plugin/FilterEntryTest.java:
```diff
@@ -145,11 +143,24 @@ public void testOverride() throws Exception {
         consumer.close();
     }
 
+    private void setMockFilterToTopic(PersistentTopic topicRef, List<EntryFilterWithClassLoader> mockFilter) throws NoSuchFieldException, IllegalAccessException {
+        Field field = topicRef.getClass().getSuperclass().getDeclaredField("entryFilters");
+        field.setAccessible(true);
+        field.set(topicRef, mockFilter);
+    }
+
+    private void setMockBrokerFilter(List<EntryFilterWithClassLoader> mockFilter) throws NoSuchFieldException, IllegalAccessException {
+        Field field2 = pulsar.getBrokerService().getEntryFilterProvider()
+                .getClass().getDeclaredField("brokerEntryFilters");
+        field2.setAccessible(true);
+        field2.set(pulsar.getBrokerService().getEntryFilterProvider(), mockFilter);
+    }
```
I hope that there's a way to get rid of these as explained in my other comment.

eolivelli reviewed (approved):
LGTM

@hangc0276 please take a look

lhotari commented:
> Unfortunately we need to keep a reference to the NarClassLoader, this is a pattern we have in other plugins that can be loaded from a Nar file

@eolivelli That can be achieved without making `EntryFilterWithClassLoader` the top level type. The top level type should be simply `EntryFilter`. It's an implementation detail that doesn't need to be exposed if there happens to be such behavior that the NarClassloader gets closed when the entry filter gets closed.

nicoloboschi commented:
@lhotari @eolivelli I've updated the pull:
- moved the internal entriesFilter update on policies change callback
- moved the top level class to `EntryFilter`
- created a mock version of `EntryFilterProvider`. unfortunately the test class is not using ´TestPulsarService´ (yet https://github.com/apache/pulsar/pull/19376) so I had to inject it in the brokerService. 

For the test part, I'm planning to improve the mocking part after https://github.com/apache/pulsar/pull/19376 gets merged, does it make sense to you @lhotari  ?

lhotari commented:
> For the test part, I'm planning to improve the mocking part after #19376 gets merged, does it make sense to you @lhotari ?

@nicoloboschi sounds good!

lhotari reviewed (approved):
Good work @nicoloboschi

nicoloboschi commented:
/pulsarbot rerun-failure-checks

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/plugin/EntryFilterProvider.java:
```diff
@@ -177,12 +180,55 @@ private static EntryFilterWithClassLoader load(EntryFilterMetaData metadata,
+    }
+
+    private static String classLoaderKey(Path archivePath) {
+        return archivePath.toFile().getAbsolutePath();
```
is archivePath.toFile().getAbsolutePath() potentially performing some IO operation ? (metadata access?)

probably archivePath.toString() is enough in this context

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/plugin/EntryFilterWithClassLoader.java:
```diff
@@ -18,12 +18,15 @@
 
 @Slf4j
+@ToString
 public class EntryFilterWithClassLoader implements EntryFilter {
```
do we still need this class ?

nicoloboschi replied:
yes, it's the implementation of EntryFilter

eolivelli commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/plugin/EntryFilterProvider.java:
```diff
@@ -37,74 +43,85 @@
+    @VisibleForTesting
+    protected Map<String, EntryFilterMetaData> definitions;
+    @VisibleForTesting
+    protected Map<String, NarClassLoader> cachedClassLoaders;
```
is it required that we refer to NarClassLoader here ? can it be simply a "ClassLoader" ?

nicoloboschi replied:
no because we need to get the ServiceDefinition file from the classloader, so a NarClassLoader is required

eolivelli commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/testcontext/MockEntryFilterProvider.java:
```diff
@@ -0,0 +1,65 @@
+            meta.setDefinition(def);
+            meta.setArchivePath(Path.of(name));
+            definitions.put(name, meta);
+            final NarClassLoader ncl = mock(NarClassLoader.class);
```
if cachedClassLoaders is not required to old " NarClassLoader" entries that maybe this code can be simplified a lot.
and you can create a new URLClassloader with Thread.currentThread().getContextClassLoader() as parent
and so we continue to remove mocks

nicoloboschi replied:
see my previous comment

eolivelli reviewed (changes requested):
Nice work

I left some additional comments

eolivelli reviewed (approved):
Lgtm
